django_pg_simple_hll/migrations/0001_initial.sql

```diff
@@ -0,0 +1,145 @@
+-- The state transition function
+-- it takes the current state, and the two input values
+-- hll_agg_state is the current running state, made up of bucket keys and hashes
+--  it is an array of n_buckets + 2
+-- input is the element we are considering
+-- hll_precision is the precision we are using for the approximation
+--  it use used to calculate the number of buckets
+create or replace function
+hll_bucket(hll_agg_state int[], input anyelement, hll_precision int)
+returns int[]
+language plpgsql immutable strict parallel safe as
+$$
+declare
+    n_buckets int := POW(2, hll_precision); -- eg.. for 512 buckets, hll_precision is 9 because 2**9 -> 512
+    hashed_input int := hashtext(input::text); -- transform the input into a text and hash it
+    bucket_hash int := hashed_input & ~(1 << 31); -- keep all the significant bits, without the sign
+    bucket_key int := hashed_input & (n_buckets - 1); -- bucket the hash into one of n buckets
+    -- pre fetch the hash for occupying the corresponding bucket for the input
+    -- we add 2 because
+    --  1. postgres arrays are 1-indexed
+    --  2. we have to account for an extra first element
+    current_hash int := hll_agg_state[bucket_key + 2];
+begin
+    -- we can only handle precision up to 11 or 2048 buckets
+    -- because hashtext produces a hash of int size with 31 useable bits
```

Comment:
This is again my naive sense of what's going on, but i would have thought the precision could go up to 31 if "hashtext produces a hash of int size with 31 useable bits"

Reply:
you are right. I think I confused an earlier attempt to make the hash 64 bits long 

I've corrected it here e305c7289c992f6d59c9c382b7ed9e20770b118d